### PR TITLE
[Diagnostics] Fix Confusing Protocol Diagnostic

### PR DESCRIPTION
**Pull request by OnyekachiSamuel**
#### What's in this pull request

- Update the diagnostic error message
- Add educational notes to explain better the error condition

Resolves [SR-13039](https://bugs.swift.org/browse/SR-13039)
Resolves: rdar://problem/64517054

### CODE REVIEW

**Comment by OnyekachiSamuel** [comment]
@hborla Kindly take a look, thanks.

File: userdocs/diagnostics/protocol-type-non-conformance.md
Comment (theblixguy) [inline userdocs/diagnostics/protocol-type-non-conformance.md:41]: I think it might be worth explaining the exceptions to the self-conformance rule - `@objc` protocols with no static requirements, `Any` and `Error`. I think a sentence or two might be enough.

File: include/swift/AST/DiagnosticsSema.def
Comment (theblixguy) [inline include/swift/AST/DiagnosticsSema.def:1800]: Three suggestions:

1) The diagnostics should always start with lowercase letter, and you don’t need “A” at the beginning here, since the diagnostics use an abbreviated form of English known as RWR (reduced written register) 🙂

2) I’m not sure if it’s worth special casing `Error` because we’d have to change the diagnostic again if Swift gains additional self-conforming protocols, but perhaps it’s okay for now.

3) The “type %1 does not conform to the protocol itself” part is only correct if both types are the same. I think you need two separate diagnostics here - one for a generic “does not conform” and one for self-conformance, refined conformance, etc.

Comment (OnyekachiSamuel) [reply include/swift/AST/DiagnosticsSema.def:1800]: @theblixguy Thanks for your feedback, I will work on them.

Comment (AnthonyLatsis) [reply include/swift/AST/DiagnosticsSema.def:1800]: I’m not sure there’s anything to gain from splitting this into several diagnostics. Perhaps we would be better off just mentioning the contrary — `; protocol types cannot conform to other existential types`— to cover abstract conformances as well.

Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def:1800]: Sure, a single diagnostic would be preferable (sometimes separate ones are easier to manage when the diagnostic text/select logic is more complicated, but it depends on what text we settle for this one).

Comment (hborla) [reply include/swift/AST/DiagnosticsSema.def:1800]: I don't think it's bad to have a _tailored_ diagnostic for the protocol type not conforming to the protocol itself - that's exactly what the SR is asking for, because reading something like `Value of protocol type 'Cancellable' cannot conform to 'Cancellable'` is confusing. We could definitely put this into a "single" diagnostic and select which text to use based on a bool like @theblixguy mentioned.

Comment (CodaFi) [reply include/swift/AST/DiagnosticsSema.def:1800]: I also recommend trying to find a way to phrase this error text in the same way you would a headline. Suppose you had code that was trying to use an existential in such a way, and try to complete the phrase "extra, extra, ____"

**Comment by varungandhi-apple** [comment]
Apologies for only mentioning `Error` as a special case in the SR. I didn't realize there were other special cases too. 😅

File: include/swift/AST/DiagnosticsSema.def
Comment (hborla) [reply include/swift/AST/DiagnosticsSema.def:1800]: Alternatively, that extra piece of information could be attached as a note. We're pretty inconsistent about when information is presented after `;` in the main error message and when it's attached as a note. In this case, I don't think it's particularly important to show that information in the main error message (but others may disagree!).

Comment (AnthonyLatsis) [reply include/swift/AST/DiagnosticsSema.def:1800]: > I don't think it's bad to have a _tailored_ diagnostic for the protocol type not conforming to the protocol itself - that's exactly what the SR is asking for, because reading something like `Value of protocol type 'Cancellable' cannot conform to 'Cancellable'` is confusing. We could definitely put this into a "single" diagnostic and select which text to use based on a bool like @theblixguy mentioned.

The self-conforming case is pretty rare, whereas `value of protocol type ‘Comparable’ does not conform to ‘Equatable’` is just as confusing unless you read the full error. I believe the extra info coming after `;` should be enough to clarify the error for all cases, and I’m worried we may end up raising more questions by unnecessarily sprinkling sugar over one specific case.

> I don't think it's particularly important to show that information in the main error message (but others may disagree!).

I would prefer to keep it in the main message for 2 reasons:
- It carries an important clarification, but Xcode doesn’t reveal attached notes inline unless you tap on the main error.
- I tend to treat notes more as representing hints or non-interactive ways to solve the error/warning being attached to.

Comment (hborla) [reply include/swift/AST/DiagnosticsSema.def:1800]: > The self-conforming case is pretty rare

I actually think this is a common misconception.

> I tend to see notes more as representing hints or non-interactive ways to resolve the error/warning they’re attached to.

I agree, and I think this note falls into that case. I think the note is hinting at passing an argument with a non-existential type, but that's not necessarily the "right" fix for the developer. For example, perhaps the developer actually wants to change whatever requires the conformance to take an existential instead.

> Xcode doesn’t reveal attached notes inline unless you tap on the main error

I definitely understand the discoverability issue with notes in Xcode, but I also don't think it's right to cram as much clarifying information as possible into the one-line error message to make up for that.

Is there a way we can clarify the main error message _and_ move the extra information about types that can conform into a note? I've been trying to avoid the term "existential" here - is that the clarifying piece that you think is missing from the first clause @AnthonyLatsis ?

Comment (hborla) [reply include/swift/AST/DiagnosticsSema.def:1800]: I'm starting to think that using the term "protocol type" to avoid "existential" is what makes this error message confusing.

Comment (varungandhi-apple) [reply include/swift/AST/DiagnosticsSema.def:1800]: The terminology used in TSPL is "protocol type". The [Protocols page](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html) has 4 mentions of "protocol type" and only one mention of "existential type". I agree with Brent's point in the SR that we should avoid using "existential" in the diagnostic if possible, but having it in the educational note would be fine IMO as we have more space to explain details.

Comment (hborla) [reply include/swift/AST/DiagnosticsSema.def:1800]: I know it's the terminology in TSPL, but in the longer form text there's more clarification about what it means to use a protocol as a type. TSPL also introduces the term "existential type". I'm not saying that we absolutely should use "existential type" here - I'm simply wondering whether saying "protocol type X cannot conform to protocol X" might be causing some confusion.

Comment (hborla) [reply include/swift/AST/DiagnosticsSema.def:1800]: What if we switched it around to say `protocol X used as a type cannot conform to protocol Y`. Is that any better? It seems to make it more clear that there's a distinction between a protocol and using a protocol as a type

Comment (varungandhi-apple) [reply include/swift/AST/DiagnosticsSema.def:1800]: > What if we switched it around to say protocol X used as a type cannot conform to protocol Y or protocol X as a type cannot conform to protocol Y. Is that any better?

I like this option better. I agree that the similarity might create more confusion, especially for non-native speakers.

File: userdocs/diagnostics/protocol-type-non-conformance.md
Comment (xwu) [inline userdocs/diagnostics/protocol-type-non-conformance.md:2]: Educational notes are written in a didactic style that start from most general to specific.

I'd suggest taking a look at the related note about [protocols with associated types](https://github.com/apple/swift/blob/373dcc76ec9968212f27c5baab49e924bcd62770/userdocs/diagnostics/associated-type-requirements.md), which starts off with an introduction that's very relevant here:

> Protocols in Swift may be used as types

Perhaps something like the following text would be helpful:

> Protocols in Swift may be used as types, but a protocol type (sometimes called an 'existential type') does not conform to its own protocol.

Because this particular question has been asked many times, you can find several different explanations on the Swift Forums. It might be helpful to see which ones seem to have the greatest clarity, and to reach out to those authors to see if they'd be willing to help contribute their work to this note.

File: userdocs/diagnostics/protocol-type-non-conformance.md
Comment (xwu) [inline userdocs/diagnostics/protocol-type-non-conformance.md:40]:
```suggestion
There is no implementation for `foo()` used above.
```

I think there's a rich set of real-world examples in the Swift Forums that users have stumbled on. Those can often be more valuable for teaching than a contrived example. I've been told that Swift documentation tries to avoid placeholder names in code examples like `i`, `x`, `foo`, instead preferring something more descriptive. Here, though, there's no semantics to be more descriptive about; I think it'd be much easier if we started from a real-world example someone has actually had trouble with, since any of those APIs would have real names and real semantics.

One simple example is probably enough, if it's well chosen. Having illustrated and described the issue, though (as I said to @owenv about the note on PATs), it can be helpful to show (at least briefly) what users who encounter this error should do _instead_. After all, they've got an error in their code, and they're here because they don't fully understand it. Here are some possible suggestions, but we should look to what the real-world examples tend to look like and try to see which ones are most frequently applicable:

In the case of an array, some users could use a homogeneous array instead; perhaps it'd have to be of type `[Any]`, though. In some cases, users could (or even should) use generics. In the general case, however, users would need to create a manual type-eraser type; _how_ to implement such a type would firmly be outside the scope of this note, however.

File: include/swift/AST/DiagnosticsSema.def
Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def:1800]: “only struct/enum/class types can conform to protocols”

I am wondering if we should also mention tuples in the educational notes, because as of SE-0283, tuples conform to Equatable, Hashable & Comparable. What do you think?

Comment (varungandhi-apple) [reply include/swift/AST/DiagnosticsSema.def:1800]: Since SE-0283 hasn't been merged yet, we can update the educational note as part of that #28833 or as a follow-up after both this PR and 28833 have been merged.

File: userdocs/diagnostics/protocol-type-non-conformance.md
Comment (OnyekachiSamuel) [reply userdocs/diagnostics/protocol-type-non-conformance.md:2]: Thanks @xwu  for the feedback, I will work on them.

File: userdocs/diagnostics/protocol-type-non-conformance.md
Comment (hborla) [inline userdocs/diagnostics/protocol-type-non-conformance.md:14]: I would suggest this be reworded slightly to make it clear that it's the implementation of these requirements that we can't access, not the protocol itself:

> Allowing existential types to conform to protocols is unsound. For protocols with static method, initializer, or associated type requirements, the implementation of these requirements cannot be accessed from the protocol type - accessing these kinds of requirements must be done using a concrete type

File: userdocs/diagnostics/protocol-type-non-conformance.md
Comment (hborla) [inline userdocs/diagnostics/protocol-type-non-conformance.md:36]: Instead of saying "write a type erasure", I would instead say "use type erasure" in the first sentence. Type erasure is a strategy that you can implement by writing a type that performs type erasure, but that "type eraser" type isn't called a type erasure. I hope that makes sense, it's a really subtle distinction.

File: userdocs/diagnostics/protocol-type-non-conformance.md
Comment (hborla) [inline userdocs/diagnostics/protocol-type-non-conformance.md:44]: `Error` is a Swift protocol, not an `@objc` protocol without static requirements. Instead of saying "One example is" you could instead say "Another exception is"

File: lib/Sema/CSDiagnostics.cpp
Comment (varungandhi-apple) [inline lib/Sema/CSDiagnostics.cpp:477]: Nit: Could you add a line-break here (and similarly hard-wrap lines in other places)? For most things, including code and documentation, we follow the [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html#source-code-width) which uses an 80-column width.

There is a tool `llvm-project/clang/tools/clang-format/clang-format-diff.py` in case you'd like to use it to format diffs, but that sometimes does too much.

File: userdocs/diagnostics/protocol-type-non-conformance.md
Comment (OnyekachiSamuel) [reply userdocs/diagnostics/protocol-type-non-conformance.md:44]: Okay, thanks @hborla  I will make the changes.

File: lib/Sema/CSDiagnostics.cpp
Comment (OnyekachiSamuel) [reply lib/Sema/CSDiagnostics.cpp:477]: Okay, thanks @varungandhi-apple, I will make the necessary changes.

**Review by hborla — APPROVED** [review]
Thank you for your work on this!

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
It looks like there were [three](https://ci.swift.org/job/swift-PR-Linux-smoke-test/23323/consoleFull) failures in the earlier test run:

```
06:37:27 Failing Tests (3):
06:37:27     Swift(linux-x86_64) :: compiler_crashers_fixed/00017-llvm-foldingset-llvm-attributesetnode-nodeequals.swift
06:37:27     Swift(linux-x86_64) :: compiler_crashers_2_fixed/0196-rdar48937223.swift
06:37:27     Swift(linux-x86_64) :: Sema/type_checker_crashers_fixed/rdar27830834.swift
```

@OnyekachiSamuel Could you fix the other two as well? They're also just diagnostic message adjustments.

**Comment by OnyekachiSamuel** [comment]
@swift-ci  please smoke test

**Comment by theblixguy** [comment]
@OnyekachiSamuel Only people with commit access can trigger CI testing. The tests are still running, but you can push new changes and one of us could trigger the tests again for you once you’re ready.

**Comment by varungandhi-apple** [comment]
I don't think you'll be able to invoke CI, since access is typically granted after a few contributions. Once you're done updating the tests, please feel free to ping whoever is reviewing (in this case, you could @ any of us), and we can kick off the CI again for you.

Also, you've probably seen this, but if you haven't, you can use the `--filter` flag (both `utils/run-test` and `lit.py` support it) to run only selected tests when you are iterating on them locally, so you don't need to run all of them.

**Comment by OnyekachiSamuel** [comment]
> I don't think you'll be able to invoke CI, since access is typically granted after a few contributions. Once you're done updating the tests, please feel free to ping whoever is reviewing (in this case, you could @ any of us), and we can kick off the CI again for you.
> 
> Also, you've probably seen this, but if you haven't, you can use the `--filter` flag (both `utils/run-test` and `lit.py` support it) to run only selected tests when you are iterating on them locally, so you don't need to run all of them.

Thanks @varungandhi-apple  for informing me, I was not aware, thanks for the tips too 👍🏻.

**Comment by OnyekachiSamuel** [comment]
> It looks like there were [three](https://ci.swift.org/job/swift-PR-Linux-smoke-test/23323/consoleFull) failures in the earlier test run:
> 
> ```
> 06:37:27 Failing Tests (3):
> 06:37:27     Swift(linux-x86_64) :: compiler_crashers_fixed/00017-llvm-foldingset-llvm-attributesetnode-nodeequals.swift
> 06:37:27     Swift(linux-x86_64) :: compiler_crashers_2_fixed/0196-rdar48937223.swift
> 06:37:27     Swift(linux-x86_64) :: Sema/type_checker_crashers_fixed/rdar27830834.swift
> ```
> 
> @OnyekachiSamuel Could you fix the other two as well? They're also just diagnostic message adjustments.

Sure, I will do that.

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by varungandhi-apple** [comment]
Merged! 🎉 ⛵

**Pull request merged**
